packages/core/lib/FileSystemCache.js

```diff
@@ -0,0 +1,97 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+const {join} = require('path');
+const {rmdir, readdir, readFile, writeFile, unlink, existsSync, mkdirSync} = require('fs');
+const {promisify} = require('util');
+const crypto = require('crypto');
+
+const path = require('path');
+const readFileAsync = promisify(readFile);
+const writeFileAsync = promisify(writeFile);
+const readdirAsync = promisify(readdir);
+const rmdirAsync = promisify(rmdir);
+const unlinkAsync = promisify(unlink);
+
+class FileSystemCache {
+  static get(
+    opts = {
+      baseDir: join(__dirname, '.cache'),
+    }
+  ) {
+    return new FileSystemCache(opts);
+  }
+
+  constructor(opts) {
+    this.opts = opts;
+    this.cache = new Map();
+  }
+
+  async get(key, defaultValue = null) {
+    let value = this.cache.get(key);
+    if (value) {
+      return value;
+    }
+    const cacheFile = this.createCacheFileName(key);
+    try {
+      const content = await readFileAsync(cacheFile, 'utf-8');
+      value = JSON.parse(content);
```

Comment:
Why must it be JSON? If you use a raw buffer, your cache becomes binary-compatible (and less error prone and less code).

Reply:
Have you got an example for this?

Reply:
```
const content = await fs.readFile(cacheFile); // returns Buffer
this.cache.set(key, value); // Buffer in Map
...
fs.writeFile(cacheFile, value); // writes Buffer to file, no need to specify encoding
```

`JSON.parse` also works on a `Buffer`, by coercing it to string first (i.e. `JSON.parse(str) === JSON.parse(Buffer.from(str))`).

Reply:
Got it, so the advantage is that you don't need to mess with encoding?